### PR TITLE
In-Timestep Remeshing

### PR DESCRIPTION
**Pull request by zfergus**
Implementation of our paper: ["In-Timestep Remeshing for Contacting Elastodynamics" [Ferguson et al. 2023]](https://zferg.us/research/in-timestep-remeshing/).

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -263,6 +257,10 @@ polyfem_target_link_system_libraries(polyfem PUBLIC natsort::natsort)
 include(glob)
 polyfem_target_link_system_libraries(polyfem PUBLIC Glob::Glob)
 
+# Tracy library
```
Comment (zfergus) [inline CMakeLists.txt:260]: Delete this

File: CMakeLists.txt [hunk 2]
```diff
@@ -286,6 +284,10 @@ if(POLYFEM_WITH_CLIPPER)
     polyfem_target_link_system_libraries(polyfem PUBLIC clipper::clipper)
     target_compile_definitions(polyfem PUBLIC -DPOLYFEM_WITH_CLIPPER)
 endif()
+# else()
```
Comment (zfergus) [inline CMakeLists.txt:287]: Add else back

File: cmake/recipes/clipper.cmake [hunk 3]
```diff
@@ -7,17 +7,14 @@ endif()
...
 )
-FetchContent_MakeAvailable(clipper_clipper)
-
-
+FetchContent_MakeAvailable(clipper)
 
-add_library(clipper_clipper ${clipper_clipper_SOURCE_DIR}/cpp/clipper.cpp)
-target_include_directories(clipper_clipper PUBLIC ${clipper_clipper_SOURCE_DIR}/cpp)
-add_library(clipper::clipper ALIAS clipper_clipper)
+add_library(clipper ${clipper_SOURCE_DIR}/cpp/clipper.cpp)
```
Comment (zfergus) [inline cmake/recipes/clipper.cmake:18]: revert this to `clipper_clipper`

File: cmake/recipes/ipc_toolkit.cmake [hunk 4]
```diff
@@ -11,7 +11,7 @@ include(FetchContent)
 FetchContent_Declare(
     ipc_toolkit
     GIT_REPOSITORY https://github.com/ipc-sim/ipc-toolkit.git
-    GIT_TAG c844956c748d6a086613f22361c5304ecaa16303
+    GIT_TAG e5dda8c173946d5f1df2a5027322cb9484504b71
```
Comment (zfergus) [inline cmake/recipes/ipc_toolkit.cmake:14]: update this to `v1.1.0`

File: cmake/recipes/wmtk.cmake [hunk 5]
```diff
@@ -11,8 +11,8 @@ message(STATUS "Third-party: creating target 'wmtk::toolkit'")
 include(FetchContent)
 FetchContent_Declare(
     wildmeshing_toolkit
-    GIT_REPOSITORY https://github.com/wildmeshing/wildmeshing-toolkit.git
-    GIT_TAG 3c2364a84613861d3314ae089abd40a54ef08929
+    GIT_REPOSITORY https://github.com/zfergus/wildmeshing-toolkit.git
```
Comment (zfergus) [inline cmake/recipes/wmtk.cmake:14]: eventually this needs to be changed to `wildmeshing`

File: input-spec.json [hunk 6]
```diff
@@ -1382,6 +1609,12 @@
         "max": 6,
         "doc": "BDF order"
     },
+    {
+        "pointer": "/time/quasistatic",
+        "type": "bool",
+        "default": false,
+        "doc": "Quasistatic time integration (i.e., zero out velocity after each time step)."
```
Comment (zfergus) [inline input-spec.json:1616]: replace `ignore_inertia` with this option

File: input-spec.json [hunk 7]
```diff
@@ -1430,7 +1663,7 @@
     },
     {
         "pointer": "/contact/use_convergent_formulation",
-        "default": false,
+        "default": true,
```
Comment (zfergus) [inline input-spec.json:1666]: revert this

File: input-spec.json [hunk 8]
```diff
@@ -3813,6 +4050,30 @@
...
+        "pointer": "/output/paraview/options/tensor_values",
+        "default": true,
+        "type": "bool",
+        "doc": "If true, write out tensor values"
+    },
+    {
+        "pointer": "/output/paraview/options/discretization_order",
+        "default": true,
+        "type": "bool",
+        "doc": "If true, write out tensor values"
```
Comment (zfergus) [inline input-spec.json:4069]: fix doc

File: input-spec.json [hunk 9]
```diff
@@ -3813,6 +4050,30 @@
...
+    },
+    {
+        "pointer": "/output/paraview/options/nodes",
+        "default": true,
+        "type": "bool",
+        "doc": "If true, write out tensor values"
```
Comment (zfergus) [inline input-spec.json:4075]: fix doc

File: input-spec.json [hunk 10]
```diff
@@ -3767,7 +4000,11 @@
             "contact_forces",
             "friction_forces",
             "velocity",
-            "acceleration"
+            "acceleration",
+            "scalar_values",
```
Comment (zfergus) [inline input-spec.json:4004]: I think these were renamed

File: src/polyfem/assembler/Assembler.cpp [hunk 11]
```diff
@@ -18,35 +18,53 @@ namespace polyfem::assembler
 		class LocalThreadMatStorage
 		{
 		public:
-			SparseMatrixCache cache;
+			std::unique_ptr<MatrixCache> cache = nullptr;
```
Comment (zfergus) [inline src/polyfem/assembler/Assembler.cpp:21]: Break this into a separate PR

File: src/polyfem/io/MshWriter.cpp [hunk 12]
```diff
@@ -53,8 +53,10 @@ namespace polyfem::io
 		mshio::MshSpec out;
 
 		auto &format = out.mesh_format;
-		format.version = "4.1";            // Only version "2.2" and "4.1" are supported.
-		format.file_type = binary ? 1 : 0; // 0: ASCII, 1: binary.
+		format.version = "4.1"; // Only version "2.2" and "4.1" are supported.
+		format.file_type = 0;   // 0: ASCII, 1: binary.
+		// NOTE: binary is broken on Greene
```
Comment (zfergus) [inline src/polyfem/io/MshWriter.cpp:58]: Revert

File: src/polyfem/io/OutData.cpp [hunk 13]
```diff
@@ -1324,6 +1329,27 @@ namespace polyfem::io
 			}
 		}
 
+		// TODO: wrap this in an if (opts.all_forces)
```
Comment (zfergus) [inline src/polyfem/io/OutData.cpp:1332]: Delete this. It is part of PR #225

File: src/polyfem/solver/NonlinearSolver.tpp [hunk 14]
```diff
@@ -164,7 +166,7 @@ namespace cppoptlib
 			// Use the maximum absolute displacement value divided by the timestep,
 			// so the units are in velocity units.
 			// TODO: Also divide by the world scale to make this criteria scale invariant.
-			this->m_current.xDelta = delta_x_norm / dt;
+			this->m_current.xDelta = descent_strategy == 2 ? NaN : (delta_x_norm / dt);
```
Comment (zfergus) [inline src/polyfem/solver/NonlinearSolver.tpp:169]: This prevents using `xDelta` if doing gradient descent.

File: src/polyfem/solver/NonlinearSolver.tpp [hunk 15]
```diff
@@ -224,8 +226,8 @@ namespace cppoptlib
 		// Log results
 		// -----------
 
-		if (this->m_status == Status::IterationLimit)
-			log_and_throw_error("[{}] Reached iteration limit (limit={})", name(), this->m_stop.iterations);
+		// if (this->m_status == Status::IterationLimit)
```
Comment (zfergus) [inline src/polyfem/solver/NonlinearSolver.tpp:229]: Renable this with a boolean flag.

File: src/polyfem/solver/SolveData.cpp [hunk 16]
```diff
@@ -156,6 +156,10 @@ namespace polyfem::solver
 		if (contact_enabled)
 		{
 			const bool use_adaptive_barrier_stiffness = !barrier_stiffness.is_number();
+			if (use_adaptive_barrier_stiffness)
```
Comment (zfergus) [inline src/polyfem/solver/SolveData.cpp:159]: Remove these

File: src/polyfem/solver/SparseNewtonDescentSolver.tpp [hunk 17]
```diff
@@ -40,10 +40,14 @@ namespace cppoptlib
 	template <typename ProblemType>
 	void SparseNewtonDescentSolver<ProblemType>::increase_descent_strategy()
 	{
-		if (this->descent_strategy == 0 || reg_weight > reg_weight_max)
-			this->descent_strategy++;
-		else
-			reg_weight = std::max(reg_weight_inc * reg_weight, reg_weight_min);
+		// if (this->descent_strategy == 1 && reg_weight < reg_weight_max)
```
Comment (zfergus) [inline src/polyfem/solver/SparseNewtonDescentSolver.tpp:43]: Revert this

File: src/polyfem/solver/forms/ContactForm.cpp [hunk 18]
```diff
@@ -106,6 +106,10 @@ namespace polyfem::solver
 		barrier_stiffness_ /= weight_;
 
 		logger().debug("adaptive barrier form stiffness {}", barrier_stiffness());
+
+		Eigen::VectorXd tmp;
```
Comment (zfergus) [inline src/polyfem/solver/forms/ContactForm.cpp:110]: Remove this

File: src/polyfem/solver/forms/ElasticForm.cpp [hunk 19]
```diff
@@ -6,6 +6,8 @@
 
 namespace polyfem::solver
 {
+	static constexpr size_t MAX_DENSE_CACHE_SIZE = 300;
```
Comment (zfergus) [inline src/polyfem/solver/forms/ElasticForm.cpp:9]: Split this into a different PR

File: src/polyfem/solver/forms/LinearForm.hpp [hunk 20]
```diff
@@ -0,0 +1,36 @@
+#pragma once
+
+#include "Form.hpp"
+
+#include <polyfem/utils/Types.hpp>
+
+namespace polyfem::solver
+{
+	/// @brief Linear form of the form cᵀx where c is a constant vector of coefficients.
+	class LinearForm : public Form
```
Comment (zfergus) [inline src/polyfem/solver/forms/LinearForm.hpp:10]: Is this used?

File: src/polyfem/solver/problems/StaticBoundaryNLProblem.hpp [hunk 21]
```diff
@@ -0,0 +1,26 @@
+#pragma once
+
+#include <polyfem/solver/NLProblem.hpp>
+
+namespace polyfem::solver
+{
+	class StaticBoundaryNLProblem : public NLProblem
```
Comment (zfergus) [inline src/polyfem/solver/problems/StaticBoundaryNLProblem.hpp:7]: Is this used?

File: src/polyfem/state/StateLoad.cpp [hunk 22]
```diff
@@ -173,4 +173,24 @@ namespace polyfem
 			}
 		}
 	}
+
+	std::unordered_map<int, std::array<bool, 3>>
+	State::boundary_conditions_ids(const std::string &bc_type) const
```
Comment (zfergus) [inline src/polyfem/state/StateLoad.cpp:178]: Is this used?

File: src/polyfem/state/StateSolveNonlinear.cpp [hunk 23]
```diff
@@ -61,29 +64,108 @@ namespace polyfem
 	{
 		init_nonlinear_tensor_solve(sol, t0 + dt);
 
-		save_timestep(t0, 0, t0, dt, sol, Eigen::MatrixXd()); // no pressure
+		const bool remesh_enabled = args["space"]["remesh"]["enabled"];
```
Comment (zfergus) [inline src/polyfem/state/StateSolveNonlinear.cpp:67]: Clean this up

File: src/polyfem/time_integrator/ImplicitTimeIntegrator.cpp [hunk 24]
```diff
@@ -55,6 +55,16 @@ namespace polyfem
 			dt_ = dt;
 		}
 
+		void ImplicitTimeIntegrator::update_quantities(const Eigen::VectorXd &x, const bool quasistatic)
```
Comment (zfergus) [inline src/polyfem/time_integrator/ImplicitTimeIntegrator.cpp:58]: Remove this

File: src/polyfem/time_integrator/ImplicitTimeIntegrator.hpp [hunk 25]
```diff
@@ -43,6 +43,11 @@ namespace polyfem::time_integrator
 		/// @param x new solution vector
 		virtual void update_quantities(const Eigen::VectorXd &x) = 0;
 
+		/// @brief Update the time integration quantities (i.e., \f$x\f$, \f$v\f$, and \f$a\f$).
+		/// @param x new solution vector
+		/// @param quasistatic whether to update quasistaticly (i.e., without updating the velocity and acceleration)
+		void update_quantities(const Eigen::VectorXd &x, const bool quasistatic);
```
Comment (zfergus) [inline src/polyfem/time_integrator/ImplicitTimeIntegrator.hpp:49]: Remove this

File: src/polyfem/utils/MaybeParallelFor.tpp [hunk 26]
```diff
@@ -20,6 +20,9 @@ namespace polyfem
 #if defined(POLYFEM_WITH_CPP_THREADS)
 			par_for(size, partial_for);
 #elif defined(POLYFEM_WITH_TBB)
+			// if (size < 100)
```
Comment (zfergus) [inline src/polyfem/utils/MaybeParallelFor.tpp:23]: delete this

File: src/polyfem/utils/Timer.hpp [hunk 27]
```diff
@@ -9,87 +9,110 @@
...
+
+		inline double getElapsedTimeInSec()
+		{
+			return m_timer.getElapsedTimeInSec();
+		}
 
-			inline const igl::Timer &igl_timer()
+		inline void log_msg()
+		{
+			const static std::string log_fmt_text =
```
Comment (zfergus) [inline src/polyfem/utils/Timer.hpp:97]: Move to class

File: src/polyfem/assembler/MassMatrixAssembler.hpp [hunk 28]
```diff
@@ -0,0 +1,61 @@
...
+#include <memory>
+
+// mass matrix assembler
+namespace polyfem::assembler
+{
+	class MassMatrixAssembler
+	{
+	public:
+		/// @brief Assembles the mass matrix.
+		/// @param[in]  is_volume True if the mesh is volumetric.
```
Comment (teseoch) [inline src/polyfem/assembler/MassMatrixAssembler.hpp:20]: do we need these?

File: src/polyfem/autogen/CMakeLists.txt [hunk 29]
```diff
@@ -1,45 +1,53 @@
...
 set(AUTOGEN_BASES
 	auto_p_bases.cpp
 	auto_p_bases.hpp
 	auto_q_bases.cpp
 	auto_q_bases.hpp
 )
 
+polyfem_set_source_group(${AUTOGEN_BASES})
+if(WIN32)
+	SET_SOURCE_FILES_PROPERTIES(${AUTOGEN_BASES} PROPERTIES COMPILE_FLAGS -Od)
```
Comment (teseoch) [inline src/polyfem/autogen/CMakeLists.txt:21]: why?

File: src/polyfem/solver/forms/ElasticForm.cpp [hunk 30]
```diff
@@ -23,12 +25,24 @@ namespace polyfem::solver
 	{
 		if (assembler_.is_linear())
 			compute_cached_stiffness();
+		if (n_bases * (is_volume ? 3 : 2) <= MAX_DENSE_CACHE_SIZE) // TODO: Expose this choice
```
Comment (teseoch) [inline src/polyfem/solver/forms/ElasticForm.cpp:28]: todo

File: src/polyfem/assembler/MassMatrixAssembler.hpp (hunk 28 above)
Comment (zfergus) [reply src/polyfem/assembler/MassMatrixAssembler.hpp:20]: Oh, right I restored this file because it was where I implemented the L2-projection matrix. I can delete the unused assemble function for the mass matrix.

File: src/polyfem/autogen/CMakeLists.txt (hunk 29 above)
Comment (zfergus) [reply src/polyfem/autogen/CMakeLists.txt:21]: I only rearranged the lines to be more blocked out. I had some compiler issue at some point with this file but a clean build fixed it. I will revert.

File: src/polyfem/solver/NonlinearSolver.tpp [hunk 31]
```diff
@@ -243,8 +245,8 @@ namespace cppoptlib
...
+		// 	log_and_throw_error("[{}] Reached iteration limit (limit={})", name(), this->m_stop.iterations);
```
Comment (zfergus) [inline src/polyfem/solver/NonlinearSolver.tpp:249]: Re-enable this

File: src/polyfem/assembler/MassMatrixAssembler.hpp (hunk 28 above)
Comment (zfergus) [reply src/polyfem/assembler/MassMatrixAssembler.hpp:20]: I am going to leave it because it would be a lot to refactor to use the new version.

**Review by teseoch — APPROVED** [review]
